packages/langium/src/workspace/document-builder.ts

```diff
@@ -114,58 +213,73 @@ export class DefaultDocumentBuilder implements DocumentBuilder {
-        });
-        const docSet = new Set([
-            ...changed,
-            ...affected,
-            // Also include all documents haven't completed the document lifecycle yet
-            ...this.langiumDocuments.all.filter(e => e.state < DocumentState.Validated)
-        ]);
-        return Array.from(docSet);
-    }
-
+    /**
+     * Build the given documents by stepping through all build phases. If a document's state indicates
+     * that a certain build phase is already done, the phase is skipped for that document.
+     */
     protected async buildDocuments(documents: LangiumDocument[], options: BuildOptions, cancelToken: CancellationToken): Promise<void> {
+        this.prepareBuild(documents, options);
         // 0. Parse content
-        //  parsing is done initially for each document, but
-        //  re-parsing after changes reported by the client might have been canceled by subsequent changes, so re-parse now
-        await this.runCancelable(documents, DocumentState.Parsed, cancelToken, doc =>
-            this.langiumDocumentFactory.update(doc)
-        );
+        await this.runCancelable(documents, DocumentState.Parsed, cancelToken, doc => {
+            this.langiumDocumentFactory.update(doc);
+        });
         // 1. Index content
         await this.runCancelable(documents, DocumentState.IndexedContent, cancelToken, doc =>
             this.indexManager.updateContent(doc, cancelToken)
         );
         // 2. Compute scopes
-        await this.runCancelable(documents, DocumentState.ComputedScopes, cancelToken, doc =>
-            this.computeScopes(doc, cancelToken)
-        );
+        await this.runCancelable(documents, DocumentState.ComputedScopes, cancelToken, async doc => {
+            const scopeComputation = this.serviceRegistry.getServices(doc.uri).references.ScopeComputation;
+            doc.precomputedScopes = await scopeComputation.computeLocalScopes(doc, cancelToken);
+        });
         // 3. Linking
-        await this.runCancelable(documents, DocumentState.Linked, cancelToken, doc =>
-            this.serviceRegistry.getServices(doc.uri).references.Linker.link(doc, cancelToken)
-        );
+        await this.runCancelable(documents, DocumentState.Linked, cancelToken, doc => {
+            const linker = this.serviceRegistry.getServices(doc.uri).references.Linker;
+            return linker.link(doc, cancelToken);
+        });
         // 4. Index references
         await this.runCancelable(documents, DocumentState.IndexedReferences, cancelToken, doc =>
             this.indexManager.updateReferences(doc, cancelToken)
         );
         // 5. Validation
-        const validateDocs = documents.filter(doc => this.shouldValidate(doc, options));
-        await this.runCancelable(validateDocs, DocumentState.Validated, cancelToken, doc =>
+        const toBeValidated = documents.filter(doc => this.shouldValidate(doc));
```

Comment:
If `shouldValidate` returns `false` the document will never reach the Validated state. This state is checked in Line 113. Does that mean that the custom implementation inside the `shouldValidate` should take care about setting the build state? If so we should add this information to the documentation.

Reply:
No, the build state is set in `runCancelable` (line 282). Custom implementations of `shouldValidate` don't need to check or modify that property.

Reply:
`toBeValidated` will not contain the document for that `shouldValidate ` returns false. For this document `runCancelable`  will not be called. Or?

Reply:
Right, so the state won't be set to Validated, but that's fine. The check you mentioned in line 113 is only for documents where validation is already partially done.